File: fm-domain/src/main/java/io/zeitwert/fm/collaboration/model/impl/AggregateWithNotesMixin.java
package io.zeitwert.fm.collaboration.model.impl;

import io.dddrive.core.ddd.model.Aggregate;
import io.zeitwert.fm.collaboration.model.ItemWithNotes;
import io.zeitwert.fm.collaboration.model.ObjNote;
import io.zeitwert.fm.collaboration.model.ObjNoteRepository;
import io.zeitwert.fm.collaboration.model.enums.CodeNoteType;

import java.time.OffsetDateTime;
import java.util.List;

import static io.dddrive.util.Invariant.requireThis;

public interface AggregateWithNotesMixin extends ItemWithNotes {

	Aggregate aggregate();

	ObjNoteRepository noteRepository();

	@Override
	default List<ObjNote> getNotes() {
		return noteRepository().getByForeignKey("relatedToId", aggregate().getId()).stream().map(it -> noteRepository().get(it)).toList();
	}

	@Override
	default ObjNote addNote(CodeNoteType noteType, Object userId) {
		ObjNote note = noteRepository().create(aggregate().getTenantId(), userId, OffsetDateTime.now());
		note.setNoteType(noteType);
		note.setRelatedToId(aggregate().getId());
		return note;
	}

	@Override
	default void removeNote(Object noteId, Object userId) {
		ObjNote note = noteRepository().load(noteId);
		requireThis(aggregate().getId().equals(note.getRelatedToId()), "Note is related to this item.");
		noteRepository().delete(note, userId, OffsetDateTime.now());
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/collaboration/model/impl/AggregateWithNotesMixin.kt
package io.zeitwert.fm.collaboration.model.impl

import dddrive.app.doc.model.Doc
import dddrive.app.obj.model.Obj
import dddrive.ddd.core.model.Aggregate
import dddrive.ddd.query.query
import io.zeitwert.fm.collaboration.model.ItemWithNotes
import io.zeitwert.fm.collaboration.model.ObjNote
import io.zeitwert.fm.collaboration.model.ObjNoteRepository
import io.zeitwert.fm.collaboration.model.enums.CodeNoteType

interface AggregateWithNotesMixin : ItemWithNotes {

	fun aggregate(): Aggregate

	fun noteRepository(): ObjNoteRepository

	override val notes: List<Any>
		get() {
			val querySpec = query {
				filter { "relatedToId" eq aggregate().id }
			}
			return noteRepository().find(querySpec)
		}

	override fun addNote(
		noteType: CodeNoteType,
		userId: Any,
	): ObjNote {
		val aggregate = aggregate()
		if (aggregate is Obj) aggregate.tenantId else (aggregate as Doc).tenantId
		val note = noteRepository().create()
		note.noteType = noteType
		note.relatedToId = aggregate.id
		return note
	}

	override fun removeNote(
		noteId: Any,
		userId: Any,
	) {
		val note = noteRepository().load(noteId)
		require(aggregate().id == note.relatedToId) { "note is related to this item." }
		noteRepository().close(note)
	}

}
